### PR TITLE
Inscription detail template + Routes + Refactoring

### LINKED ISSUE #59: i0 appended twice to url

### LINKED ISSUE #97: /inscription/:id page
Ability to display similar information like at
https://ordinals.com/inscription/6fb976ab49dcec017f1e201e84395983204ae1a7c2abf7ced0a85d692e442799i0

For unconfirmed inscriptions, there will not be an inscription number just yet.

### PR DESCRIPTION
**Pull request by sectore**
- [x] Introduce `/` + `inscription/{id}` routes - contents goes into `app/routes`
- [x] Extract `Home` content from `App`
- [x] Extract + Rename `StreamValues` to `StreamContext`
- [x] Extract server functions from `App` into `app/functions/*`
- [x] Remove deprecated `count*` stuff (based on Leptos counter example)
- [x] Footer: Use `StreamContext` instead of passing props down
- [ ] ~~Add fetch inscriptions logic into `app/functions/api` incl. `Inscription` data structure~~ **Moved to** #109
- [x] Style `Inscription` component
- [x] Show preview in a layer (on-top)
- [x] Update styles in header + footer 
- [x] Fix #59 in 53aacfb 
- [x] Optimize release build by following [Appendix: Optimizing WASM Binary Size](https://leptos-rs.github.io/leptos/appendix_binary_size.html#things-to-do) and adding options for release build (66949ae)

## Preview

[Screencast from 26.04.2023 19:42:43.webm](https://user-images.githubusercontent.com/47693/234660150-e8131016-ef18-49a2-a00d-ebef500a519a.webm)

Closes #97
Fixes #59

### CODE REVIEW

**Comment by FelixWeis** [comment]
first glance looks great! refactoring aproach is welcomed. 
will have a deeper look tomorrow.
at current branch, image cycling on regtest after `just p` seems broken but for mempool infoline and block height display works fine.

**Comment by FelixWeis** [comment]
compare the behaviour of running

- [x] `git checkout <>`
- [x] `just install`
- [x] `just run-services`
- [x] `just watch`
- [ ] `just p`

cycling on main works but this PR doesn't display the new mempool inscriptions yet.

**Comment by FelixWeis** [comment]
ping me when i1p works again, thanks!

File: src/app/components/livegrid.rs [hunk 1]
```diff
@@ -3,8 +3,32 @@ use leptos::*;
...
+pub struct Item {
+  pub key: usize,
+  pub hash: ReadSignal<String>,
+  set_hash: WriteSignal<String>,
+}
+
+#[allow(dead_code)]
+impl Item {
+  pub fn new(cx: Scope, key: usize, hash: String) -> Self {
+    let (hash, set_hash) = create_signal(cx, hash);
```
Comment (sectore) [inline src/app/components/livegrid.rs:17]: Tried `create_rw_signal` first and switched back to `create_signal` afterwards - both could not fix current issues we have for updating values in `For` component.

File: src/app/components/livegrid.rs [hunk 2]
```diff
@@ -51,9 +88,15 @@ pub fn LiveGrid(
...
-              view! { cx, <InscriptionItem id=inscription_id/> }
+          key=|item| item.key.clone()
+          view=move |cx, item| {
+              let hash = move || item.hash.get();
+              view! { cx,
+                <li class="relative">
+                  <p class="text-white">{&hash()}</p>
+                  <p class="text-white">{format!("/inscription/{}", & hash())}</p>
+                </li>
+              }
```
Comment (sectore) [inline src/app/components/livegrid.rs:99]: Current issue: By updating values of `item.hash` no further rendering happened in `For` component or (with current code) it panics with `already mutably borrowed: BorrowError', `

File: src/app/routes/inscription.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                        </div>
+                      </button>
+                    </div>
+                    <Container>
+                      <Title>"id"</Title>
+                      <Label>{&id()}</Label>
+                    </Container>
+                    <Container>
+                      <Title>"address"</Title>
+                      <Label>"bc1pyqz4gyra5cssfgawk6dunh4s9car9zrzpkw2l0sz4g9ym0cdxzescnf4u2"</Label>
```
Comment (FelixWeis) [inline src/app/routes/inscription.rs:80]: Lets hide the dummy data based on `LEPTOS_ENV` or so. 
this allows us to merge this major refactor for devs without showing false info in prod.

for dummy data use the first mainnet inscription
https://ordinals.com/inscriptions/0
https://ordinals.com/inscription/6fb976ab49dcec017f1e201e84395983204ae1a7c2abf7ced0a85d692e442799i0

File: src/app/routes/home.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  let initial_inscriptions: Vec<_> = (0..6).map(|n| format!("punk_{}.webp", n)).collect();
+
+  view! { cx,
+    <div class="content py-8">
+      <div class="flex">
+        <h1 class="flex-1 text-2xl font-bold text-gray-900 dark:text-gray-100">
+          "Live unconfirmed inscriptions"
+        </h1>
+      </div>
+      <div class="text-xs text-gray-900 dark:text-gray-100">{info}</div>
```
Comment (FelixWeis) [inline src/app/routes/home.rs:20]: min-height or dummy `&nbsp;` so layout doesn't jump

**Review by FelixWeis — CHANGES_REQUESTED** [review]
Almost ready IMO. A few nits commented above.

In general we should be mindful of the change in size to our production `.wasm`

As is, the PR increases the footprint ~23% / +140KB. which seems high. Not sure if we compile so many more crates into it now with the routing or what else could be the reason. Maybe its premature optimization tho. Thoughts?

Some tools like [twiggy](https://github.com/rustwasm/twiggy) or [wabt](https://github.com/WebAssembly/wabt) might come in handy with profiling.
 
```patch
--- tmp/main.dust.txt 2023-05-02 07:20:42
+++ tmp/pr.dust.txt 2023-05-02 07:21:29
@@ -3,7 +3,7 @@
 4.0K       ┌── inline0.js                    │                          █ │   1%
 4.0K     ┌─┴ leptos_reactive-5561073d4cc9ec2d│                          █ │   1%
 4.0K   ┌─┴ snippets                          │                          █ │   1%
- 16K   ├── ordilabs_live.css                 │                          █ │   3%
- 48K   ├── ordilabs_live.js                  │                        ███ │   8%
-528K   ├── ordilabs_live.wasm                │   ████████████████████████ │  89%
-596K ┌─┴ pkg                                 │███████████████████████████ │ 100%
+ 20K   ├── ordilabs_live.css                 │                          █ │   3%
+ 48K   ├── ordilabs_live.js                  │                         ██ │   7%
+664K   ├── ordilabs_live.wasm                │  █████████████████████████ │  90%
+736K ┌─┴ pkg                                 │███████████████████████████ │ 100%
```

File: src/app/routes/inscription.rs (hunk 3 above)
Comment (sectore) [reply src/app/routes/inscription.rs]: > Lets hide the dummy data based on LEPTOS_ENV

Ok, hided in `DEV` now (2e1ef53)

> for dummy data use the first mainnet inscription

Why? Mock data are still mock data.

**Comment by sectore** [comment]
> In general we should be mindful of the change in size to our production `.wasm`
> 
> As is, the PR increases the footprint ~23% / +140KB. which seems high. Not sure if we compile so many more crates into it now with the routing or what else could be the reason. Maybe its premature optimization tho. Thoughts?
> 

Interesting. Can't tell you the WHY, this PR does not introduce any new library or something - more research is needed (later). However, by following [Appendix: Optimizing WASM Binary Size](https://leptos-rs.github.io/leptos/appendix_binary_size.html#things-to-do) and adding options for release build in 66949ae , the wasm size could be reduced to ~500kb.

Screenshot from 2023-05-02 12-21-48

File: src/app/routes/inscription.rs (hunk 3 above)
Comment (FelixWeis) [reply src/app/routes/inscription.rs]: as an homage to casey

**Comment by FelixWeis** [comment]
Looks good, hiding the dummy info works, altough I haven't figured out the exact combination 
```toml
#Cargo.toml
[package.metadata.leptos]
env = "DEV" # PROD
```

```bash
# serve release
LEPTOS_ENV=prod cargo serve -r
```

when using `-r` or `--release` it seems to trigger a similar bug we currently have in [staging](https://s-live.ordilabs.org/) where it seems to append a second body instead of replacing it.

**Comment by FelixWeis** [comment]
ACK 1f347927ac0921e5ea085d5a0eb18c151a882cf8

**Pull request merged**